scipy/sparse/csgraph/_matching.pyx

```diff
@@ -235,41 +233,40 @@ cdef tuple _hopcroft_karp(ITYPE_t[:] indices, ITYPE_t[:] indptr,
                 # Initialize stack to contain only w and reset path.
                 stack[0] = w
                 stack_head = 1
-                path_head = 0
                 while stack_head != 0:
-                    # Pop v from stack and push it to path.
+                    # Pop v from stack.
                     stack_head -= 1
                     v = stack[stack_head]
-                    path[path_head] = v
-                    path_head += 1
                     could_augment = False
                     for up in range(indptr[v], indptr[v + 1]):
                         u = indices[up]
-                        if dist[y[u]] == dist[v] + 1:
+                        yu = y[u]
+                        if dist[yu] == dist[v] + 1:
                             could_augment = True
-                            # If y[u] is unmatched, we have found an augmenting
+                            # If yu is unmatched, we have found an augmenting
                             # path. We update the matching and move on to the
                             # next unmatched vertex.
-                            if y[u] == i:
+                            if yu == i:
                                 done = True
-                                while path_head != 0:
-                                    path_head -= 1
-                                    v = path[path_head]
+                                # Unwind and follow the path back to the root.
+                                while True:
```

Comment:
Thanks for the review! I think I did consider using something akin to `v != w` but didn't as it would become a bit wonky: At least, we can't just change the condition to `v != w` and get rid of the check later, since then we wouldn't be performing the augmentation at the root node. Conversely, `parents[w]` isn't defined, so that needs to go after the check.

Reply:
You're right. And I'm not inclined to make a working algorithm more complex - the intent is also already explicitly commented and explained in the above comment.